envoy/network/socket.h

```diff
@@ -92,6 +93,18 @@ class SocketAddressProvider {
    * @param indent_level the level of indentation.
    */
   virtual void dumpState(std::ostream& os, int indent_level) const PURE;
+
+  /**
+   * @return the upstream SSL connection. This will be nullptr if the upstream
+   * connection does not use SSL.
+   */
+  virtual Ssl::ConnectionInfoConstSharedPtr upstreamSslConnection() const PURE;
```

Comment:
I think this interface should only have the downstreamSslConnection(). All the other fields here are specific to the downstream connection, so including the upstream connection seems misleading.

Reply:
Yes, I feel strange too. Although we can keep this in the StreamInfo also, but it still feel strange, since the StreamInfo is only mean to stream for one direction(upstream, or downstream). Also can see the downstream StreamInfo's upstream SSL info is copied from upstream StreamInfo. For example: 
https://github.com/envoyproxy/envoy/blob/a8033fa8e44406ba6448cf3110037cb79f81ff39/source/common/tcp_proxy/tcp_proxy.cc#L531

So in the end, the upstream SSL info is still mixed with downstream SSL info. Based on my guess the reason is the filter need the upstream StreamInfo, but other than tcpProxy or HttpRouter filter can't get the upstream StreamInfo, since the upstream StreamInfo is created the last filter (tcpProxy/http router filter).

Reply:
The streaminfo is a bit odd, and has a mix of upstream/downstream. But `SocketAddressProvider` is only supposed to have information that should be copied to each request on the connection. The upstream ssl info should not be copied between requests, so I don't think it belongs here.

Reply:
thanks, got it, let me revert the change to upstream SSL info.